cardano-node/src/Cardano/Tracing/OrphanInstances/Shelley.hs

```diff
@@ -1026,6 +1026,7 @@ instance ToJSON (Alonzo.CollectError StandardCrypto) where
               Alonzo.InlineDatumsNotSupported -> String "Inline datums not supported"
               Alonzo.ReferenceScriptsNotSupported -> String "Reference scripts not supported"
               Alonzo.ReferenceInputsNotSupported -> String "Reference inputs not supported"
+              Alonzo.TimeTranslationPastHorizon -> String "Time translation requested past the horizon"
```

Comment:
Where can I find more information about the "horizon"? I think we can improve this error message.

Reply:
1) I'm unaware of user facing documentation for this. 😬 

2) There's a whole chapter of the Consensus Report about it "Part IV Hard Fork Combinator", "Chapter 17 Time", where "beyond the horizon" means "past the Forecast range". But it's very abstract/fundamental/not necessarily in "Cardano terminology".

3) High-level summary (@JaredCorduan please review this):

* the horizon is either the end of the current epoch or the end of the next epoch
    * it's the end of the current epoch if the ledger state does not determine whether the current epoch is the end of the current era
    * it's the end of the next epoch if the ledger state does determine whether the current epoch is the end of the current era
    * (ie the horizon is the end of the current epoch if and only if we don't know whether it's the last epoch of this era)
    
* when does the ledger state does determine whether the current epoch is the end of the current era?

    * the transition happens at the end of the current epoch if and only if all the necessary proposal/voting transactions are in a block on the chain too deep to be rolled back _and_ there's a stability window's worth of slots between that block and the end of the current epoch.
    * The current epoch is determined as the last of its era when the above is already true in the ledger state.
    * The current epoch is determined as _not_ the last of its era when the above _cannot_ be true, eg it's not true now and we're already within a stability window of the end of the epoch.
    * Otherwise, it's not determined: ie it's not true now but it still might become true.

Reply:
That all sounds correct to me, though I might re-phrase it a bit. I would say that the ledger is ignorant about transitions to future eras, and that the consensus layer uses the ledger state to determine if an era is ending.

In particular, votes to change the protocol parameters in the next epoch _must_ occur within two stability windows of the end of the epoch (the vote states what epoch it is intended for, you can also place a pending vote for the following epoch). After one stability window from the end of the epoch (so one stability window _after_ voting has ended), the consensus layer looks to see if the major protocol version (which is a protocol parameter). is going to change, and hence usher in a new era (provided it is not an intra-era hard fork).

Reply:
I will come back to this another time, thanks for the input